sdk/assemblyscript/src/models/openai/chat.ts

```diff
@@ -1538,3 +1539,51 @@ export class AudioOutput {
    */
   transcript!: string;
 }
+
+/**
+ * Represents a a raw message object, which is used when reconstructing messages from JSON.
+ * A raw message will round-trip all the JSON data, but does not expose the fields directly.
+ * (note, this type is not exported)
+ */
+class RawMessage extends RequestMessage {
+  constructor(data: string) {
+    const obj = JSON.parse<JSON.Obj>(data);
+    if (!obj.has("role")) {
+      throw new Error("Missing role field in message JSON.");
+    }
+
+    const role = obj.get("role")!.get<string>();
+    super(role);
+
+    this._data = data;
+  }
+
+  private _data: string;
+
+
+  @serializer
+  serialize(self: RawMessage): string {
+    return self._data;
+  }
+
+
+  @deserializer
+  deserialize(data: string): RawMessage {
+    return new RawMessage(data);
+  }
+}
+
+/**
+ * Parses a JSON-encoded request message into a list of RequestMessage objects.
+ * The resulting message objects are suitable for restoring a previous chat conversation.
+ * However, the original message types are not preserved.
+ */
+export function parseMessages(data: string): RequestMessage[] {
```

Comment:
This can be slightly simplified using `>=json-as@1.0.8`. Changing from `JSON.Value[]` to `JSON.Obj[]` also ensures that `data` can only be an array of `objects` instead of accepting all types.

```js
export function parseMessages(data: string): RequestMessage[] {
  return JSON.parse<JSON.Obj[]>(data).map<RequestMessage>(msg => new RawMessage(msg.toString()));
}
```

Reply:
I tried that, but got the same result.